packages/react/src/utils/createBaseUIEventData.ts

```diff
@@ -0,0 +1,63 @@
+import type { PopupChangeReason } from './types';
+
+/**
+ * Maps an open-change `reason` string to the corresponding native event type.
+ */
+export type ReasonToEvent<Reason extends string> = Reason extends 'trigger-press'
+  ? MouseEvent | PointerEvent | TouchEvent | KeyboardEvent
+  : Reason extends 'trigger-hover'
+    ? MouseEvent
+    : Reason extends 'outside-press'
+      ? MouseEvent | PointerEvent
+      : Reason extends 'item-press' | 'close-press'
+        ? MouseEvent | KeyboardEvent | PointerEvent
+        : Reason extends 'cancel-open'
+          ? MouseEvent
+          : Reason extends 'trigger-focus' | 'focus-out'
+            ? FocusEvent
+            : Reason extends 'escape-key' | 'list-navigation'
+              ? KeyboardEvent
+              : Event;
+
+/**
+ * Discriminated union keyed by the `reason` string literal.
+ * Narrows `event` type based on `reason`.
+ */
+export type BaseUIEventData<Reason extends string = PopupChangeReason> = {
+  [K in Reason]: {
```

Comment:
Add JSDocs to properties. Also, this type should be exported from the package.

Reply:
We're not AFAIK. Let's create a top-level `types.ts` file and import the type from there. Then add a /types export path in package.json.